### PR TITLE
Unify SGMCMC init structure

### PR DESCRIPTION
**Pull request by SamDuffield**
As discussed in #537 , this PR unifies the `sgmcmc` module such that all algorithms have an `init` function and return a `MCMCSamplingAlgorithm`, in the case of `sgld` and `sghmc` the init function is the trivial `lambda position: position`.

This does not result in totally seamless switching since some algorithms (`sgnht` and `csgld`) require a `rng_key` whereas others do not. However, the block to seamless switching is not just down to the `init` since some algorithms (`csgld`) also require custom `step_fun` signatures.

I have reordered the `init` signatures such that `position` is first, `rng_key` is second (if required) followed by any other 
algorithm-specific arguments. This included a refactor to `ghmc` and as a results `meads_adaptation`.

I suggest that possibly we should consider a global reordering of `step_fn` with `state` before `rng_key` since it is possible for an algorithm to not require a `rng_key` in their `step_fn`, e.g. SVGD #512

Finally, I unified the `temperature` parameter such that it is given as an input to the `step_fn` with a default value of 1, previously it had a default value of 0.01 for `csgld` which I am not sure was intended.

I also left the `init` out of the `docs` for `sgld` and `sghmc` since they still work without it, i.e. only using `.step`

 A few important guidelines and requirements before we can merge your PR:

 - [x] We should be able to understand what the PR does from its title only;
 - [x] There is a high-level description of the changes;
 - [x] There are links to *all* the relevant issues, discussions and PRs;
 - [x] The branch is rebased on the latest `main` commit;
 - [x] Commit messages follow these [guidelines](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html);
 - [x] The code respects the current naming conventions;
 - [x] Docstrings follow the [numpy style guide](https://numpydoc.readthedocs.io/en/latest/format.html)
 - [x] `pre-commit` is installed and configured on your machine, and you ran it before opening the PR;
 - [x] There are tests covering the changes;
 - [x] The doc is up-to-date;

### CODE REVIEW

**Comment by junpenglao** [comment]
Is it possible to not use `rng_key` in `init`? For example, could we replace it with a placeholder value/PyTree? Would replacing the  `rng_key` with a fix key works?

**Comment by SamDuffield** [comment]
> Is it possible to not use `rng_key` in `init`? For example, could we replace it with a placeholder value/PyTree? Would replacing the `rng_key` with a fix key works?

Yep! We could get close to seamless switching by either
1. Enforce a `rng_key` placeholder even for algorithms that do not require an `init`
```
alg = blackjax.sgld
# or e.g. alg = blackjax.sgnht

state = alg.init(position, rng_key1)
state = alg.step(rng_key2, state, minibatch)
```
2. Set a default fixed rng_key 
```
alg = blackjax.sgld
# or e.g. alg = blackjax.sgnht

state = alg.init(position)
state = alg.step(rng_key, state, minibatch)
```

Although both options have drawbacks (1 clutters the API of those algorithms that do not require `rng_key` on `init`, 2 uses a default fixed key which could hide the pseudo-randomness of the algorithm initiation).

Note that this would allow seamless switching for `sgld`, `sghmc` and `sgnht` but not for `csgld` since `csgld` has a different `step` signature with `step_size_diff` and `step_size_stoch`

File: blackjax/sgmcmc/csgld.py [hunk 1]
```diff
@@ -216,11 +228,12 @@ def init_fn(position: PyTree):
 
         def step_fn(
             rng_key: PRNGKey,
-            state,
+            state: ContourSGLDState,
             minibatch: PyTree,
             step_size_diff: float,
             step_size_stoch: float,
-        ):
+            temperature: float = 1.0,
```
Comment (albcab) [inline blackjax/sgmcmc/csgld.py:235]: In Appendix J of https://arxiv.org/pdf/1902.03932.pdf they do some analysis on the robustness of the temperature, concluding that "We find that the optimal temperature is often less than 1. We hypothesize that this result is due to the model misspecification common to neural networks." For CIFAR-10 the optimal temperature is 0.1 and for CIFAR-100 it is 0.01. So I guess this is why they chose it, but not sure what would be the best default value in this case.

File: blackjax/sgmcmc/sgnht.py [hunk 2]
```diff
@@ -127,19 +127,25 @@ class sgnht:
...
+            position: PyTree, rng_key: PRNGKey, init_xi: Union[None, float] = None
+        ):
+            return cls.init(position, rng_key, init_xi or alpha)
+
+        def step_fn(
+            rng_key: PRNGKey,
+            state: SGNHTState,
+            minibatch: PyTree,
+            step_size: float,
+            temperature: float = 1,
```
Comment (albcab) [inline blackjax/sgmcmc/sgnht.py:145]: As I understand it, SGNHT controls the temperature with `xi`, so we shouldn't need a temperature other than 1. If this is the case, we should at least add a `*,` before temperature so that the user that changes it knows what she is doing.

**Review by albcab — APPROVED** [review]
> I suggest that possibly we should consider a global reordering of step_fn with state before rng_key since it is possible for an algorithm to not require a rng_key in their step_fn, e.g. SVGD https://github.com/blackjax-devs/blackjax/pull/512

Considering #541 this makes sense, some sampling algorithms have deterministic steps. Most sampling algorithms have random steps though, so maybe we keep the default as is for now.

> Although both options have drawbacks (1 clutters the API of those algorithms that do not require rng_key on init, 2 uses a default fixed key which could hide the pseudo-randomness of the algorithm initiation).

Samplers that require a `rng_key` at `init` have some persistent value, like a momentum. The initialization of this value could impact the resulting samples so I think it's important to make this explicit. Because sampling algorithms don't usually have persistent values, I think the default should remain as is.

File: blackjax/sgmcmc/csgld.py (hunk 1 above)
Comment (SamDuffield) [reply blackjax/sgmcmc/csgld.py:235]: I see ok, yes I think for our use case a default value of 1. makes more sense since we do not assume the model context never mind any cold posterior effect, and this is in line with the other `sgmcmc` algorithms

File: tests/mcmc/test_sampling.py [hunk 3]
```diff
@@ -270,7 +270,7 @@ def test_linear_regression_sgld(self):
 
         data_batch = X_data[:100, :]
         init_position = 1.0
-        _ = sgld(rng_key, init_position, data_batch, 1e-3)
+        _ = sgld.step(rng_key, init_position, data_batch, 1e-3)
```
Comment (albcab) [inline tests/mcmc/test_sampling.py:273]: Also what about adding `init_position = sgld.init(init_position)` to keep the coverage up?

File: blackjax/sgmcmc/sgnht.py (hunk 2 above)
Comment (SamDuffield) [reply blackjax/sgmcmc/sgnht.py:145]: I think this might be one case where the terminology is a bit confusing 😅 , my understanding from [Ding et al](https://proceedings.neurips.cc/paper/2014/file/21fe5b8ba755eeaece7a450849876228-Paper.pdf) is that `xi` automatically modifies the friction coefficient to force the "system temperature" to be near a "target temperature". This target temperature is the one defined above with default value 1, and behaves in the same way as in the other `sgmcmc` algorithms (just scales the logpdf/gradient).

Comment (SamDuffield) [reply blackjax/sgmcmc/sgnht.py:145]: Actually looking at it now I think the temperature scaling might be incorrect in `diffusions.sghmc` and `diffusions.sgnht`

It should be either simply rescaling the gradient `g / temperature` or equivalently all `stepsize * temperature` right? Atm for `diffusions.sghmc` and `diffusions.sgnht` I think its doing the latter but incorrectly

Comment (albcab) [reply blackjax/sgmcmc/sgnht.py:145]: I see what you mean. Change was made in #396. It feels like `g / temperature` is the best option here.

Comment (SamDuffield) [reply blackjax/sgmcmc/sgnht.py:145]: I think I actually prefer the `step_size * temperature` parameterization as it results in a nice transition to optimisation algorithms with `temperature` -> 0
SGLD -> SGD
SGHMC -> SGD with momentum
SGNHT -> something funky/cool/new 👀 

For SGHMC and SGNHT the temperature is applied to the augmented target distribution $\pi(x, v) \propto \exp(- (U(x) + \frac12 v^Tv) / T)$ rather than just the target in $x$

Probably this should be explained in the docs, and the transition to gradient descent(s) might make a nice little example.

All that needs to be changed is `step_size * beta` -> `step_size * temperature * beta` in `sghmc` and `sgnht` - I'll update the PR shortly.

**Comment by SamDuffield** [comment]
Added `sgld.init` and `sghmc.init` to the tests and amended the temperature scaling for `sghmc` and `sgnht`
